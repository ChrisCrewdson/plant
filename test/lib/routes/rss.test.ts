const bson = require('bson');

const { ObjectID } = bson;

/** @type {BizPlant[]} */
const plants = [{
  _id: 'pid2',
  locationId: 'lid1',
  userId: 'uid1',
  title: 'Test Title',
}];

const testObjectId = new ObjectID('012345678901234567890123');
const _id = testObjectId;
const userId = testObjectId;
const plantIds = [testObjectId];
/** @type {DbNoteWithPlants[]} */
const dbNoteWithPlants = [{
  _id,
  date: 20170606,
  plants,
  plantIds,
  userId,
}];

const mockMongo = () => ({
  getNotesLatest: async () => Promise.resolve(dbNoteWithPlants),
});

jest.mock('../../../lib/db/mongo', () => mockMongo);

const rss = require('../../../lib/routes/rss.js');

const req = {
  protocol: 'fake protocol',
  get: () => 'fake host',
};

const res = {
  status: () => res,
  /** @param {string} builtXml */
  send: (_: string) => {}, // eslint-disable-line no-unused-vars
};

describe('rss', () => {
  test('renders xml', (done) => {
    /** @param {string} builtXml */
    res.send = (builtXml) => {
      expect(builtXml).toMatchSnapshot();
      done();
    };
    /** @type {import("express").Application} */
    const app = {
      // @ts-ignore - ignore for testing
      get: (_, routeAction) => {
        routeAction(req, res);
      },
    };
    rss(app);
  });

  // Use different input values to force target code down
  // different paths and increase coverage.
  test('renders xml on alternate paths', (done) => {
    plants[0].title = '';
    plants[1] = plants[0]; // eslint-disable-line prefer-destructuring
    // @ts-ignore - can be removed once this is correctly typed
    dbNoteWithPlants[0].images = [{
      id: '',
      ext: '.png',
      originalname: 'orig-file-name',
      size: 1234,
      sizes: [{
        name: 'orig',
        width: 1000,
      }],
    }];
    // @ts-ignore - can be removed once this is correctly typed
    dbNoteWithPlants[0].note = 'Some text';
    /** @param {string} builtXml */
    res.send = (builtXml) => {
      expect(builtXml).toMatchSnapshot();
      done();
    };
    /** @type {import("express").Application} */
    const app = {
      // @ts-ignore - ignore for testing
      get: (_, routeAction) => {
        routeAction(req, res);
      },
    };
    rss(app);
  });
});
