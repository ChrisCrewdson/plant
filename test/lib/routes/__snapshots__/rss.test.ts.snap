// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rss renders xml 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<rss version=\\"2.0\\" xmlns:atom=\\"http://www.w3.org/2005/Atom\\">
  <channel>
    <title>Latest Notes</title>
    <atom:link href=\\"fake protocol://fake host/rss\\" rel=\\"self\\" type=\\"application/rssxml\\" />
    <language>en-us</language>
    <link>fake protocol://fake host/rss</link>
    <description>Plant progress.</description>
    <item>
        <title>Test Title</title>
        <description><![CDATA[(no comment)]]></description>
        <pubDate>Tue, 06 Jun 2017 00:00:00 PST</pubDate>
        <guid>fake protocol://fake host/plant/test-title/012345678901234567890123?noteid=012345678901234567890123#012345678901234567890123</guid>
        <link>fake protocol://fake host/plant/test-title/012345678901234567890123?noteid=012345678901234567890123#012345678901234567890123</link>
      </item>
  </channel>
</rss>"
`;

exports[`rss renders xml on alternate paths 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<rss version=\\"2.0\\" xmlns:atom=\\"http://www.w3.org/2005/Atom\\">
  <channel>
    <title>Latest Notes</title>
    <atom:link href=\\"fake protocol://fake host/rss\\" rel=\\"self\\" type=\\"application/rssxml\\" />
    <language>en-us</language>
    <link>fake protocol://fake host/rss</link>
    <description>Plant progress.</description>
    <item>
        <title>(none) (+ 1 others)</title>
        <description><![CDATA[<p>Some text</p>
<img src=\\"https://i.plaaant.com/up/orig/..png\\" style=\\"max-width: 100%\\" />]]></description>
        <pubDate>Tue, 06 Jun 2017 00:00:00 PST</pubDate>
        <guid>fake protocol://fake host/plant/none/012345678901234567890123?noteid=012345678901234567890123#012345678901234567890123</guid>
        <link>fake protocol://fake host/plant/none/012345678901234567890123?noteid=012345678901234567890123#012345678901234567890123</link>
      </item>
  </channel>
</rss>"
`;
